src/packages/client/src/__tests__/integration/happy/long-running-transaction/test.ts

```diff
@@ -0,0 +1,374 @@
+import { getTestClient } from '../../../../utils/getTestClient'
+
+let PrismaClient, prisma
+
+describe('long-running transaction', () => {
+  /**
+   * Minimal example of a long-running transaction
+   */
+  test('basic', async () => {
+    const result = await prisma.$transaction(async (prisma) => {
+      await prisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      })
+
+      await prisma.user.create({
+        data: {
+          email: 'user_2@website.com',
+        },
+      })
+
+      return prisma.user.findMany()
+    })
+
+    expect(result.length).toBe(2)
+  })
+
+  /**
+   * Transactions should fail after the default timeout
+   */
+  test('timeout default', async () => {
+    const result = prisma.$transaction(async (prisma) => {
+      await prisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      })
+
+      await new Promise((res) => setTimeout(res, 6000))
+
+      return prisma.user.findMany()
+    })
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
+
+            Invalid \`prisma.user.findMany()\` invocation:
+
+
+              Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Expired'.
+          `)
+  })
+
+  /**
+   * Transactions should fail after the changed timeout
+   */
+  test('timeout override', async () => {
+    const result = prisma.$transaction(
+      async (prisma) => {
+        await prisma.user.create({
+          data: {
+            email: 'user_1@website.com',
+          },
+        })
+
+        await new Promise((res) => setTimeout(res, 600))
+
+        return prisma.user.findMany()
+      },
+      {
+        maxWait: 200,
+        timeout: 500,
+      },
+    )
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
+
+            Invalid \`prisma.user.findMany()\` invocation:
+
+
+              Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Expired'.
+          `)
+  })
+
+  /**
+   * Transactions should fail and rollback if thrown within
```

Comment:
Would need a test for the case where the commit or rollback fails.

Reply:
There is one `already committed`, is that what you were thinking of?

Reply:
Not really, it needs to test the case of a network failure or delay on the commit (aka retry logic I was talking about)